### PR TITLE
Add manufactured solution tendency terms to Omega

### PR DESCRIPTION
**Pull request by hyungyukang**
This PR adds custom tendency terms required for the manufactured solution test case to Omega. This feature can be tested in conjunction with the Polaris package, which has included a convergence study leveraging this capability. The details of this test case can be found in Section 2.10 and Figures 13 and 19 in [Bishnu et al. 2024](https://doi.org/10.1029/2022MS003545) and the [manufactured solution test](https://docs.e3sm.org/polaris/main/users_guide/ocean/tasks/manufactured_solution.html) in Polaris.

The error norms for this test case are expected to demonstrate second-order convergence. The test results will be posted in this PR following integration with Polaris for this test case.

This PR does not modify the unit tests but passed them on Frontier using both CPU and GPU.

Checklist
* [ ] Testing
  * [ ] Unit tests have passed. Please provide a relevant CDash build entry for verification.
  * [ ] Polaris test suite has passed

### CODE REVIEW

**Comment by hyungyukang** [comment]
Although personal testing indicates second-order convergence for 'Default,' 'Del2,' and 'Del4,' this must be validated with Polaris. 

image

The above test was conducted on Frontier GPU using crayclanggpu.

File: components/omega/src/ocn/Tendencies.cpp
Comment (xylar) [inline components/omega/src/ocn/Tendencies.cpp:48]: It isn't clear to me why this isn't a critical error:
```suggestion
   Err = TendConfig.get("UseCustomTendency", UseCustomTendency);
   if (Err != 0) {
      LOG_CRITICAL("Tendencies: UseCustomTendency not found in Config");
   }
```
If we aren't going to initialize tendencies, I think we should stop the code.

Comment (xylar) [inline components/omega/src/ocn/Tendencies.cpp:48]: It isn't clear to me why this isn't a critical error:
```suggestion
   if (Err != 0) {
      LOG_CRITICAL("Tendencies: UseCustomTendency not found in Config");
   }
```
If we aren't going to initialize tendencies, I think we should stop the code.

File: components/omega/src/ocn/Tendencies.cpp
Comment (xylar) [inline components/omega/src/ocn/Tendencies.cpp:62]: Presumably, we will never return this value because we will have quit the code?

File: components/omega/src/ocn/Tendencies.cpp
Comment (xylar) [inline components/omega/src/ocn/Tendencies.cpp:72]: Same comment.

File: components/omega/src/ocn/Tendencies.cpp
Comment (xylar) [inline components/omega/src/ocn/Tendencies.cpp:87]: I realize this is not really relevant to this PR but a question to developers more broadly: Should we check the error here and call `LOG_CRITICAL()` if there was a problem?

**Review by xylar — COMMENTED** [review]
A few comments/questions about error checking.  These are not just for @hyungyukang but for Omega developers more broadly.  Maybe @philipwjones has some thoughts?

**Comment by mwarusz** [comment]
@hyungyukang
>  May I include your name in the reviewer list if you are comfortable with it?

Sure.

**Comment by hyungyukang** [comment]
@xylar and @cbegeman , 

To enable this feature, `omega.yml` requires some changes from `config/Default.yml`.
Here is the difference between `omega.yml` I used and `config/Default.yml`.


<details open>
<summary>Diff. between `omega.yml` and `config/Default.yml`</summary>
<br>

```diff
@@ -1,13 +1,13 @@
 Omega:
   TimeIntegration:
     CalendarType: No Leap
     TimeStepper: Forward-Backward
     TimeStepper: RungeKutta4
     TimeStep: 0000_00:10:00
     StartTime: 0001-01-01_00:00:00
-    StopTime: 0001-01-01_02:00:00
+    StopTime: 0001-01-01_10:00:00
     RunDuration: none
   Dimension:
-    NVertLevels: 60
+    NVertLevels: 1
   Decomp:
     HaloWidth: 3
     DecompMethod: MetisKWay
@@ -30,8 +30,8 @@ Omega:
     EddyDiff2: 10.0
     TracerHyperDiffTendencyEnable: true
     EddyDiff4: 0.0
-    UseCustomTendency: false
-    ManufacturedSolutionTendency: false
+    UseCustomTendency: true
+    ManufacturedSolutionTendency: true
   Tracers:
     Base: [Temperature, Salinity]
     Debug: [Debug1, Debug2, Debug3]
@@ -49,7 +49,6 @@ Omega:
       UseStartEnd: false
       Contents:
         - State
-        - Base
     # Restarts are used to initialize for all job submissions after the very
     # first startup job. We use UseStartEnd with a start time just after the
     # simulation start time so that omega does not attempt to use a restart
@@ -84,8 +83,8 @@ Omega:
       Mode: write
       IfExists: replace
       Precision: double
-      Freq: 1
-      FreqUnits: months
+      Freq: 10
+      FreqUnits: hours
       UseStartEnd: false
       Contents:
         - Tracers
```

</details>

**Comment by philipwjones** [comment]
@xylar and all. As I mentioned in the Error design, most of these errors should indeed be critical and abort, but we don't have a way of doing that easily until I get the new Error capability implemented (LOG_CRITICAL currently does not abort automatically). As part of that implementation, I'll be doing a pass through the code and replacing many of these error calls and error returns with critical errors and aborts when needed.  But if we want to at least upgrade these to critical error messages, that will be a good hint about where we want this to happen.

**Comment by philipwjones** [comment]
And while I'm commenting on this PR...do we need to start thinking about multiple YAML configs now that we're starting to get conflicts. In particular, the streams unit test with this change will be writing output every 10 hours for a full year so probably not what we want.

**Comment by xylar** [comment]
@philipwjones, thanks for the update.  I didn't realize that we weren't there yet with LOG_CRITICAL, sorry about that.

**Comment by xylar** [comment]
>  In particular, the streams unit test with this change will be writing output every 10 hours for a full year so probably not what we want.

I don't think @hyungyukang is proposing any changes to the YAML file (other than adding a few config options).  These are the changes that Polaris will be making to the YAML file it uses (made manually for now).

**Comment by philipwjones** [comment]
Ah yes, I misread the comment on the Default.yml changes.

**Comment by xylar** [comment]
We need a Polaris PR to make the changes to the manufactured solution tests such that the correct yaml config options are set.  This shouldn't be hard but there are a lot of moving pieces in Polaris and Omega right now so I haven't got to it yet.

File: components/omega/src/ocn/CustomTendencyTerms.cpp
Comment (mwarusz) [inline components/omega/src/ocn/CustomTendencyTerms.cpp:95]: This is outside the scope of this PR, but should we start thinking about creating a shared constants file ? @mark-petersen @xylar Thoughts ?

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (mwarusz) [inline components/omega/src/ocn/CustomTendencyTerms.h:56]:
```suggestion
```
`XCell` and `YCell` are local variables already.

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (mwarusz) [inline components/omega/src/ocn/CustomTendencyTerms.h:42]: This function doesn't need to be defined in the header file. Its implementation can be moved to the cpp file. Also should probably have a `const` annotation:  

```suggestion
      void operator()(Array2DReal ThicknessTend, const OceanState *State,
                      const AuxiliaryState *AuxState, int ThickTimeLevel,
                      int VelTimeLevel, TimeInstant Time)  const {
```

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (mwarusz) [inline components/omega/src/ocn/CustomTendencyTerms.h:97]: Same as above, this function doesn't need to be defined in the header file. Its implementation can be moved to the cpp file. Also should probably have a `const` annotation:

```suggestion
      void operator()(Array2DReal NormalVelTend, const OceanState *State,
                      const AuxiliaryState *AuxState, int ThickTimeLevel,
                      int VelTimeLevel, TimeInstant Time) const {
```

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (mwarusz) [inline components/omega/src/ocn/CustomTendencyTerms.h:115]:
```suggestion
```

**Review by mwarusz — COMMENTED** [review]
I have a few C++ coding suggestions and one general question about constants.

File: components/omega/src/ocn/CustomTendencyTerms.cpp
Comment (xylar) [reply components/omega/src/ocn/CustomTendencyTerms.cpp:95]: These need (!!!) to come from:
https://github.com/E3SM-Project/E3SM/blob/09b9b7e21e3d381eafa962e178a5951580e2fc74/share/util/shr_const_mod.F90

How is Scream handling this?

Comment (mwarusz) [reply components/omega/src/ocn/CustomTendencyTerms.cpp:95]: Not sure that they do:

https://github.com/E3SM-Project/E3SM/blob/09b9b7e21e3d381eafa962e178a5951580e2fc74/components/eamxx/src/physics/share/physics_constants.hpp#L46

https://github.com/E3SM-Project/E3SM/blob/09b9b7e21e3d381eafa962e178a5951580e2fc74/share/util/shr_const_mod.F90#L16

**Review by xylar — APPROVED** [review]
@hyungyukang, this is working great for me in Polaris. On Chrysalis, I did a test merge of this branch with `develop`.  Then, ran the CTest utility in Polaris, which is also a convenient way to automatically build the Omega code.  Then, I ran the manufactured solution test using this branch:
https://github.com/E3SM-Project/polaris/pull/249

The results look great!

convergence_ssh

comparison

For everyone's reference, a typical `omega.yml` that Polaris produces for this is:
``` yaml
Omega:
  TimeIntegration:
    CalendarType: No Leap
    TimeStepper: RungeKutta4
    TimeStep: 0000_00:10:00.000
    StartTime: 0001-01-01_00:00:00
    StopTime: none
    RunDuration: 0000_10:00:00.000
  Dimension:
    NVertLevels: 1
  Decomp:
    HaloWidth: 3
    DecompMethod: MetisKWay
  State:
    NTimeLevels: 2
  Advection:
    FluxThicknessType: Center
    FluxTracerType: Center
  Tendencies:
    ThicknessFluxTendencyEnable: true
    PVTendencyEnable: true
    KETendencyEnable: true
    SSHTendencyEnable: true
    VelDiffTendencyEnable: false
    ViscDel2: 1.0e3
    VelHyperDiffTendencyEnable: false
    ViscDel4: 1.2e11
    TracerHorzAdvTendencyEnable: true
    TracerDiffTendencyEnable: true
    EddyDiff2: 10.0
    TracerHyperDiffTendencyEnable: true
    EddyDiff4: 0.0
    UseCustomTendency: true
    ManufacturedSolutionTendency: true
  Tracers:
    Base: [Temperature, Salinity]
    Debug: [Debug1, Debug2, Debug3]
  ManufacturedSolution:
    WavelengthX: 5.0e6
    WavelengthY: 4.33013e6
    Amplitude: 1.0
  IOStreams:
    # InitialState should only be used when starting from scratch
    # After the simulations initial start, the frequency should be
    # changed to never so that the initial state file is not read.
    InitialState:
      UsePointerFile: false
      Filename: init.nc
      Mode: read
      Precision: double
      Freq: 1
      FreqUnits: OnStartup
      UseStartEnd: false
      Contents:
      - NormalVelocity
      - LayerThickness
    RestartRead:
      UsePointerFile: true
      PointerFilename: ocn.pointer
      Mode: read
      Precision: double
      Freq: 1
      FreqUnits: OnStartup
      UseStartEnd: true
      StartTime: 0001-01-01_00:00:01
      EndTime: 99999-12-31_00:00:00
      Contents:
      - Restart
    History:
      UsePointerFile: false
      Filename: output.nc
      Mode: write
      IfExists: replace
      Precision: double
      Freq: 36000
      FreqUnits: Seconds
      UseStartEnd: false
      Contents:
      - NormalVelocity
      - LayerThickness
      - SshCellDefault
```

**Comment by hyungyukang** [comment]
@xylar , this is great! Thank you so much for testing this PR with Polaris! Results look beautiful. Were you able to run Del2/Del4 test as well?

**Comment by xylar** [comment]
@hyungyukang, I think some work will be needed by @cbegeman to rebase https://github.com/E3SM-Project/polaris/pull/234 before we can run del2/del4 in Polaris.  She's busy with AGU right now and I haven't had time to familiarize myself with that branch yet.  So I didn't try that so far.  Maybe we should treat that as a separate review?

**Comment by hyungyukang** [comment]
@xylar , Aha, got it. Probably we can wait for it? I'm currently busy with AGU as well and will need some time to address the code suggestions from @mwarusz . While I don't want to hold this PR for longer time, it will not affect subsequent Omega developments. So it would be good to merge it once all tests (default, del2, del4) have been confirmed to work properly.

**Comment by xylar** [comment]
> So it would be good to merge it once all tests (default, del2, del4) have been confirmed to work properly.

We have a bit of a logjam at the moment.  Manufactured solution default is the only test case that supports Omega at the moment so it would be really nice to have it working with both Omega `develop` and Polaris `main`.  This would require merging this PR, updating the Omega submodule in https://github.com/E3SM-Project/polaris/pull/231 and merging it, and then merging https://github.com/E3SM-Project/polaris/pull/249.

If https://github.com/E3SM-Project/polaris/pull/234 can be rebased and updated to work with the convergence in space and time only that we recently added, we should be able to test with it.  But I don't feel comfortable taking over that work so that will need to wait for @cbegeman to have some time.  My preference would be to merge this branch assuming that the del2 and del4 terms are okay based on your testing, and to follow up with another PR if we later discover that bug fixes are needed.

File: components/omega/src/ocn/CustomTendencyTerms.cpp
Comment (mark-petersen) [reply components/omega/src/ocn/CustomTendencyTerms.cpp:95]: Agreed. Common constants need to come from a unified file, and preferably the centralized one that @xylar pointed to above. We struggled with multiple constants files in MPAS within E3SM.

**Comment by mark-petersen** [comment]
I am testing this PR on Perlmutter. CPU compiles. GPU with `nvidia` has the error within `ManufacturedThicknessTendency`, so it is caused by this PR:
```
/global/homes/m/mpeterse/repos/E3SM/opr/components/omega/src/ocn/CustomTendencyTerms.h(65): error: An extended __host__ __device__ lambda cannot be defined inside a class ("ManufacturedThicknessTendency") with private or protected access within another class
               [=] __attribute__((host)) __attribute__((device))(int ICell, int KLevel) {
                                                                ^

/global/homes/m/mpeterse/repos/E3SM/opr/components/omega/src/ocn/CustomTendencyTerms.h(133): error: An extended __host__ __device__ lambda cannot be defined inside a class ("ManufacturedVelocityTendency") with private or protected access within another class
               [=] __attribute__((host)) __attribute__((device))(int IEdge, int KLevel) {
                                                                ^
```

<details>
<summary>This is my compile sequence: </summary> <p>

```
######### perlmutter CPU
CODE_DIR=opr

cd /global/homes/m/mpeterse/repos/E3SM/${CODE_DIR}
git submodule update --init --recursive externals/ekat externals/scorpio cime
cd components/omega/

module load cmake
RUNDIR=241205_omega_cpu
mkdir ${PSCRATCH}/runs/$RUNDIR
cd !$

rm -rf build
mkdir build
cd build
export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-cpu/spack/dev_polaris_0_4_0_nvidia_mpich/var/spack/environments/dev_polaris_0_4_0_nvidia_mpich/.spack-env/view

cmake \
   -DOMEGA_CIME_COMPILER=nvidia \
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=pm-cpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /global/homes/m/mpeterse/repos/E3SM/${CODE_DIR}/components/omega -B .
./omega_build.sh

# linking:
cd test
ln -isf /global/homes/m/mpeterse/meshes/omega/O*nc .
cp /global/homes/m/mpeterse/repos/E3SM/${CODE_DIR}/components/omega/configs/Default.yml omega.yml

# run test:
salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu --account=m4572 # or e3sm
RUNDIR=241205_omega_cpu
cd ${PSCRATCH}/runs/${RUNDIR}/build

./omega_ctest.sh
```
</details>

**Comment by mark-petersen** [comment]
Also with `-DOMEGA_CIME_COMPILER=gnugpu`
```
/global/homes/m/mpeterse/repos/E3SM/opr/components/omega/src/ocn/CustomTendencyTerms.h(65): error: An extended __host__ __device__ lambda cannot be defined inside a class ("ManufacturedThicknessTendency") with private or protected access within another class
                           (int ICell, int KLevel) {
                           ^

/global/homes/m/mpeterse/repos/E3SM/opr/components/omega/src/ocn/CustomTendencyTerms.h(133): error: An extended __host__ __device__ lambda cannot be defined inside a class ("ManufacturedVelocityTendency") with private or protected access within another class
                           (int IEdge, int KLevel) {
                           ^

/global/homes/m/mpeterse/repos/E3SM/opr/components/omega/src/ocn/CustomTendencyTerms.h(65): error: An extended __host__ __device__ lambda cannot be defined inside a class ("ManufacturedThicknessTendency") with private or protected access within another class
                           (int ICell, int KLevel) {
                           ^

/global/homes/m/mpeterse/repos/E3SM/opr/components/omega/src/ocn/CustomTendencyTerms.h(133): error: An extended __host__ __device__ lambda cannot be defined inside a class ("ManufacturedVelocityTendency") with private or protected access within another class
                           (int IEdge, int KLevel) {
                           ^
```

**Comment by hyungyukang** [comment]
> In other words, while leaving this PR unmerged is not holding up Omega development, it is holding up Polaris development.

@xylar , all make sense to me. Thanks! I will address suggestions and issues soon.

**Comment by hyungyukang** [comment]
> I am testing this PR on Perlmutter. CPU compiles. GPU with `nvidiagpu` has the error within `ManufacturedThicknessTendency`, so it is caused by this PR:

@mark-petersen , thanks for testing. I actually haven't ran it on pm-gpu because sometimes compilation on it is very slow... I will address the issue soon.

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (hyungyukang) [reply components/omega/src/ocn/CustomTendencyTerms.h]: Done.

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (hyungyukang) [reply components/omega/src/ocn/CustomTendencyTerms.h]: I moved the implementation of the customized tendencies to cpp file (`CustomTendencyTerms.cpp`) and added `const` as well.

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (hyungyukang) [reply components/omega/src/ocn/CustomTendencyTerms.h]: Done.

File: components/omega/src/ocn/CustomTendencyTerms.h
Comment (hyungyukang) [reply components/omega/src/ocn/CustomTendencyTerms.h]: Done.

**Comment by hyungyukang** [comment]
> Passes perlmutter cpu tests using `-DOMEGA_CIME_COMPILER=nvidia`, except the `REDUCTIONS_TEST` which is a known problem on the head of `develop`.

@mark-petersen , I have resolved the issue on `pm-gpu`. Could you please try again?

**Review by mark-petersen — APPROVED** [review]
Yes, this compiles on pm gpu with nvidia now. Approving based on the reported tests above. Thanks!

**Comment by mark-petersen** [comment]
Checked with @cbegeman, who is busy at AGU this week.

**Pull request merged**
